compiler/src/linking/link.re

```diff
@@ -0,0 +1,529 @@
+            let value =
+              switch (imported_name) {
+              | "relocBase" =>
+                Expression.const(
+                  wasm_mod,
+                  Literal.int32(Int32.of_int(table_offset^)),
+                )
+              | "moduleRuntimeId" =>
+                incr(module_id);
+                Expression.const(
+                  wasm_mod,
+                  Literal.int32(Int32.of_int(module_id^)),
+                );
+              | _ => assert(false)
+              };
+            ignore @@
+            Global.add_global(linked_mod, new_name, Type.int32, false, value);
+          } else {
+            let ty = Global.get_type(global);
+            let mut = Global.is_mutable(global);
+            Import.add_global_import(
+              linked_mod,
+              new_name,
+              imported_module,
+              imported_name,
+              ty,
+              mut,
+            );
+          };
+        };
+      } else {
+        let internal_name = Global.get_name(global);
+        let new_name = gensym(internal_name);
+        Hashtbl.add(local_names, internal_name, new_name);
+
+        let ty = Global.get_type(global);
+        let mut = Global.is_mutable(global);
+        let init = Global.get_init_expr(global);
+
+        globalize_names(local_names, init);
+        ignore @@ Global.add_global(linked_mod, new_name, ty, mut, init);
+      };
+    };
+
+    let num_functions = Function.get_num_functions(wasm_mod);
+    let funcs =
+      List.init(num_functions, Function.get_function_by_index(wasm_mod));
+    let (imported_funcs, funcs) =
+      List.partition(is_function_imported, funcs);
+
+    List.iter(
+      func => {
+        let imported_module = Import.function_import_get_module(func);
+        if (is_grain_module(imported_module)) {
+          let imported_name = Import.function_import_get_base(func);
+          let internal_name = Function.get_name(func);
+          let new_name =
+            Hashtbl.find(
+              Hashtbl.find(exported_names, imported_module),
+              imported_name,
+            );
+          Hashtbl.add(local_names, internal_name, new_name);
+        } else {
+          let imported_name = Import.function_import_get_base(func);
+          let internal_name = Function.get_name(func);
+          let new_name = gensym(internal_name);
+          Hashtbl.add(local_names, internal_name, new_name);
+
+          let params = Function.get_params(func);
+          let results = Function.get_results(func);
+          Import.add_function_import(
+            linked_mod,
+            new_name,
+            imported_module,
+            imported_name,
+            params,
+            results,
+          );
+        };
+      },
+      imported_funcs,
+    );
+
+    List.iter(
+      func => {
+        let internal_name = Function.get_name(func);
+        let new_name = gensym(internal_name);
+        Hashtbl.add(local_names, internal_name, new_name);
```

Comment:
Whenever you gensym a new name, you need to add it to the local_names hashtbl, right? Maybe those should be bundled so it isn't missed during a code cleanup or something.

Reply:
Not always, no. I suppose we could have some utility that does this for where we _do_ need to add it to the table.